internal/alb/sg/security_group.go

```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return err
+	}
+	if instance != nil {
+		return controller.ec2.DeleteSecurityGroupByID(*instance.GroupId)
+	}
+	return nil
+}
+
+func (controller *securityGroupController) reconcileByNewSGInstance(group *SecurityGroup) error {
+	// TODO: move these VPC calls into controller startup, and part of controller configuration
+	vpcID, err := controller.ec2.GetVPCID()
+	if err != nil {
+		return err
+	}
+	createSGOutput, err := controller.ec2.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
+		VpcId:       vpcID,
+		GroupName:   group.GroupName,
+		Description: aws.String("Instance SecurityGroup created by alb-ingress-controller"),
+	})
+	if err != nil {
+		return err
+	}
+	group.GroupID = createSGOutput.GroupId
+
+	_, err = controller.ec2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+		GroupId:       group.GroupID,
+		IpPermissions: group.InboundPermissions,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = controller.ec2.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{group.GroupID},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: group.GroupName,
+			},
+			{
+				Key:   aws.String(albec2.ManagedByKey),
+				Value: aws.String(albec2.ManagedByValue),
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (controller *securityGroupController) reconcileByModifySGInstance(group *SecurityGroup, instance *ec2.SecurityGroup) error {
+	if group.GroupID == nil {
+		group.GroupID = instance.GroupId
+	}
+	if group.GroupName == nil {
+		group.GroupName = instance.GroupName
+	}
+
+	permissionsToRevoke := diffIPPermissions(instance.IpPermissions, group.InboundPermissions)
+	if len(permissionsToRevoke) != 0 {
+		_, err := controller.ec2.RevokeSecurityGroupIngress(&ec2.RevokeSecurityGroupIngressInput{
+			GroupId:       group.GroupID,
+			IpPermissions: permissionsToRevoke,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to revoke inbound permissions due to %s", err.Error())
+		}
+	}
+
+	permissionsToGrant := diffIPPermissions(group.InboundPermissions, instance.IpPermissions)
+	if len(permissionsToGrant) != 0 {
+		_, err := controller.ec2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+			GroupId:       group.GroupID,
+			IpPermissions: permissionsToGrant,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to grant inbound permissions due to:%s", err.Error())
+		}
+	}
+	return nil
+}
+
+// findExistingSGInstance tring to find the existing SG matches the specification
+func (controller *securityGroupController) findExistingSGInstance(group *SecurityGroup) (*ec2.SecurityGroup, error) {
+	if group.GroupID != nil {
+		instance, err := controller.ec2.GetSecurityGroupByID(*group.GroupID)
+		if err != nil {
+			return nil, err
+		}
+		if instance == nil {
+			return nil, fmt.Errorf("securityGroup %s doesn't exist", *group.GroupID)
+		}
+		return instance, nil
+	} else if group.GroupName != nil {
```

Comment:
No need `else` here? Just `if group.GroupName != nil {` in the next line?

Reply:
i explicitly used else if to express the idea that one of the branch should occur. (to avoid cases like someone modified the if groupId != nil branch, and assigned groupName there)

Reply:
i have used switch statements to get that point across

Reply:
changed to use switch instead